File: x-finance-scraper/scrape.py
"""
Pobiera posty z X (Twitter) na temat firmy/hasla powiazanego z finansami
z ostatnich N godzin (domyslnie 72h), dzialajac przez przegladarke
(Playwright + prawdziwy DOM x.com), a NIE przez oficjalne API.

Wymaga wczesniejszego jednorazowego logowania: `python login.py`
(zapisuje sesje do auth/storage_state.json).

Uzycie:
    python scrape.py
    python scrape.py --query "Google" --hours 72
    python scrape.py --query "Google" --keywords "finance,earnings,stock" --lang en
    python scrape.py --headful          (pokaz okno przegladarki, jesli jest ekran)
    python scrape.py --strict           (odrzucaj posty bez slowa kluczowego z finansow)
"""
import argparse
import csv
import json
import os
import random
import sys
from datetime import datetime, timedelta, timezone
from urllib.parse import quote

from playwright.sync_api import sync_playwright

ROOT = os.path.dirname(os.path.abspath(__file__))
STORAGE_STATE_PATH = os.path.join(ROOT, "auth", "storage_state.json")
OUTPUT_DIR = os.path.join(ROOT, "output")
DEFAULT_CONFIG_PATH = os.path.join(ROOT, "config", "default.json")

# Wykonywane w kontekscie strony (JS) - odczytuje dane z kazdej karty posta.
EXTRACT_JS = r"""
(articles) => articles.map((article) => {
    const timeEl = article.querySelector('time');
    const datetime = timeEl ? timeEl.getAttribute('datetime') : null;
    const linkEl = timeEl ? timeEl.closest('a') : null;
    const url = linkEl ? linkEl.href : null;

    const textEl = article.querySelector('[data-testid="tweetText"]');
    const text = textEl ? textEl.innerText.replace(/\s+/g, ' ').trim() : '';

    let handle = null;
    let displayName = null;
    const userNameBlock = article.querySelector('[data-testid="User-Name"]');
    if (userNameBlock) {
        const links = userNameBlock.querySelectorAll('a[href^="/"]');
        for (const a of links) {
            const href = a.getAttribute('href') || '';
            if (/^\/[A-Za-z0-9_]+$/.test(href)) {
                handle = href.slice(1);
                break;
            }
        }
        displayName = (userNameBlock.innerText.split('\n')[0] || '').trim() || null;
    }

    const metrics = {};
    ['reply', 'retweet', 'like'].forEach((key) => {
        const el = article.querySelector(`[data-testid="${key}"]`);
        metrics[key] = el ? (el.getAttribute('aria-label') || el.textContent || '').trim() : null;
    });

    return { url, datetime, text, handle, displayName, metrics };
})
"""


def parse_args():
    parser = argparse.ArgumentParser(description="Scraper postow z X (Twitter) przez przegladarke.")
    parser.add_argument("--query", type=str, help="Glowne haslo/firma do wyszukania")
    parser.add_argument("--keywords", type=str, help="Slowa kluczowe oddzielone przecinkami (OR)")
    parser.add_argument("--hours", type=float, help="Ile godzin wstecz pobierac posty")
    parser.add_argument("--lang", type=str, help="Kod jezyka X (np. pl, en)")
    parser.add_argument("--max-tweets", type=int, help="Limit liczby zebranych postow")
    parser.add_argument("--max-scrolls", type=int, help="Limit liczby scrolli strony wynikow")
    parser.add_argument("--strict", action="store_true", help="Post musi zawierac haslo ORAZ slowo finansowe")
    parser.add_argument("--no-exclude-replies", action="store_true", help="Nie wykluczaj odpowiedzi (replies)")
    parser.add_argument("--headful", action="store_true", help="Pokaz okno przegladarki")
    return parser.parse_args()


def load_config(args):
    with open(DEFAULT_CONFIG_PATH, "r", encoding="utf-8") as f:
        cfg = json.load(f)

    if args.query:
        cfg["query"] = args.query
    if args.keywords:
        cfg["keywords"] = [k.strip() for k in args.keywords.split(",") if k.strip()]
    if args.hours is not None:
        cfg["hoursBack"] = args.hours
    if args.lang:
        cfg["lang"] = args.lang
    if args.max_tweets:
        cfg["maxTweets"] = args.max_tweets
    if args.max_scrolls:
        cfg["maxScrolls"] = args.max_scrolls
    if args.strict:
        cfg["strictKeywordFilter"] = True
    if args.no_exclude_replies:
        cfg["excludeReplies"] = False

    cfg["headless"] = not args.headful
    return cfg


def build_search_query(cfg):
    parts = [cfg["query"]]
    keywords = cfg.get("keywords") or []
    if keywords:
        or_clause = " OR ".join(f'"{k}"' if " " in k else k for k in keywords)
        parts.append(f"({or_clause})")
    if cfg.get("excludeReplies"):
        parts.append("-filter:replies")
    if cfg.get("lang"):
        parts.append(f"lang:{cfg['lang']}")
    return " ".join(parts)


def matches_keywords(text, cfg):
    lower = text.lower()
    if not cfg.get("strictKeywordFilter"):
        return True
    query_hit = cfg["query"].lower() in lower
    kw_hit = any(k.lower() in lower for k in cfg.get("keywords", []))
    return query_hit and kw_hit


def parse_dt(iso_str):
    return datetime.fromisoformat(iso_str.replace("Z", "+00:00"))


def write_csv(path, rows):
    headers = ["datetime", "handle", "displayName", "text", "url", "reply", "retweet", "like"]
    with open(path, "w", newline="", encoding="utf-8") as f:
        writer = csv.writer(f)
        writer.writerow(headers)
        for r in rows:
            metrics = r.get("metrics") or {}
            writer.writerow([
                r.get("datetime"),
                r.get("handle"),
                r.get("displayName"),
                r.get("text"),
                r.get("url"),
                metrics.get("reply"),
                metrics.get("retweet"),
                metrics.get("like"),
            ])


def main():
    args = parse_args()
    cfg = load_config(args)

    if not os.path.exists(STORAGE_STATE_PATH):
        print(
            f"Brak pliku sesji: {STORAGE_STATE_PATH}\n"
            "Najpierw zaloguj sie jednorazowo lokalnie: python login.py",
            file=sys.stderr,
        )
        sys.exit(1)

    os.makedirs(OUTPUT_DIR, exist_ok=True)

    cutoff = datetime.now(timezone.utc) - timedelta(hours=cfg["hoursBack"])
    search_query = build_search_query(cfg)
    url = f"https://x.com/search?q={quote(search_query)}&src=typed_query&f=live"

    print(f"Zapytanie wyszukiwania: {search_query}")
    print(f"Okno czasowe: ostatnie {cfg['hoursBack']}h (od {cutoff.isoformat()})")

    seen = {}
    stale_scrolls = 0

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=cfg["headless"])
        context = browser.new_context(
            storage_state=STORAGE_STATE_PATH,
            viewport={"width": 1280, "height": 1000},
            user_agent=(
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
                "(KHTML, like Gecko) Chrome/124.0 Safari/537.36"
            ),
        )
        page = context.new_page()
        page.goto(url, wait_until="domcontentloaded")

        if "/login" in page.url or "/i/flow/login" in page.url:
            print("Sesja wygasla lub jest nieprawidlowa - zaloguj sie ponownie: python login.py", file=sys.stderr)
            browser.close()
            sys.exit(1)

        try:
            page.wait_for_selector('article[data-testid="tweet"]', timeout=20000)
        except Exception:
            print("Nie znaleziono zadnych postow dla tego zapytania (lub strona nie zaladowala sie poprawnie).")

        for i in range(cfg["maxScrolls"]):
            batch = page.eval_on_selector_all('article[data-testid="tweet"]', EXTRACT_JS)

            for t in batch:
                if not t.get("url") or not t.get("datetime"):
                    continue
                if t["url"] in seen:
                    continue
                if not matches_keywords(t.get("text", ""), cfg):
                    continue
                seen[t["url"]] = t

            times = [parse_dt(t["datetime"]) for t in batch if t.get("datetime")]
            oldest_in_batch = min(times) if times else None

            if oldest_in_batch is not None and oldest_in_batch < cutoff:
                stale_scrolls += 1
            else:
                stale_scrolls = 0

            print(f"Scroll {i + 1}/{cfg['maxScrolls']} - zebrano lacznie: {len(seen)}")

            if stale_scrolls >= 3:
                print("Kolejne posty sa juz starsze niz zadane okno czasowe - konczenie.")
                break
            if len(seen) >= cfg["maxTweets"]:
                print("Osiagnieto limit maxTweets - konczenie.")
                break

            page.mouse.wheel(0, 2200)
            page.wait_for_timeout(1200 + random.random() * 1300)

        browser.close()

    results = [t for t in seen.values() if parse_dt(t["datetime"]) >= cutoff]
    results.sort(key=lambda t: parse_dt(t["datetime"]), reverse=True)

    stamp = datetime.now().strftime("%Y-%m-%dT%H-%M-%S")
    json_path = os.path.join(OUTPUT_DIR, f"x-posts-{stamp}.json")
    csv_path = os.path.join(OUTPUT_DIR, f"x-posts-{stamp}.csv")

    with open(json_path, "w", encoding="utf-8") as f:
        json.dump(results, f, ensure_ascii=False, indent=2)
    write_csv(csv_path, results)

    print(f"\nZnaleziono {len(results)} postow z ostatnich {cfg['hoursBack']}h.")
    print(f"Zapisano do:\n  {json_path}\n  {csv_path}")


if __name__ == "__main__":
    main()

File: x-finance-scraper/src/scrape.js
/**
 * Pobiera posty z X (Twitter) na temat firmy/hasla powiazanego z finansami
 * z ostatnich N godzin (domyslnie 72h), dzialajac przez przegladarke
 * (Playwright + prawdziwy DOM x.com), a NIE przez oficjalne API.
 *
 * Wymaga wczesniejszego jednorazowego logowania: `npm run login`
 * (zapisuje sesje do auth/storageState.json).
 *
 * Uzycie:
 *   node src/scrape.js
 *   node src/scrape.js --query "Google" --hours 72
 *   node src/scrape.js --query "Google" --keywords "finance,earnings,stock" --lang en
 *   node src/scrape.js --headful          (pokaz okno przegladarki, jesli jest ekran)
 *   node src/scrape.js --strict           (odrzucaj posty bez slowa kluczowego z finansow)
 */
const { chromium } = require('playwright');
const path = require('path');
const fs = require('fs');

const ROOT = path.join(__dirname, '..');
const STORAGE_STATE_PATH = path.join(ROOT, 'auth', 'storageState.json');
const OUTPUT_DIR = path.join(ROOT, 'output');
const DEFAULT_CONFIG_PATH = path.join(ROOT, 'config', 'default.json');

function parseArgs(argv) {
  const args = {};
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (!a.startsWith('--')) continue;
    const key = a.slice(2);
    const next = argv[i + 1];
    if (next === undefined || next.startsWith('--')) {
      args[key] = true;
    } else {
      args[key] = next;
      i++;
    }
  }
  return args;
}

function loadConfig(args) {
  const base = JSON.parse(fs.readFileSync(DEFAULT_CONFIG_PATH, 'utf8'));
  const cfg = { ...base };

  if (args.query) cfg.query = String(args.query);
  if (args.keywords) cfg.keywords = String(args.keywords).split(',').map((s) => s.trim()).filter(Boolean);
  if (args.hours) cfg.hoursBack = Number(args.hours);
  if (args.lang) cfg.lang = String(args.lang);
  if (args['max-tweets']) cfg.maxTweets = Number(args['max-tweets']);
  if (args['max-scrolls']) cfg.maxScrolls = Number(args['max-scrolls']);
  if (args.strict) cfg.strictKeywordFilter = true;
  if (args['no-exclude-replies']) cfg.excludeReplies = false;

  cfg.headless = !args.headful;
  return cfg;
}

function buildSearchQuery(cfg) {
  const parts = [];
  parts.push(cfg.query);
  if (Array.isArray(cfg.keywords) && cfg.keywords.length > 0) {
    const orClause = cfg.keywords.map((k) => (k.includes(' ') ? `"${k}"` : k)).join(' OR ');
    parts.push(`(${orClause})`);
  }
  if (cfg.excludeReplies) parts.push('-filter:replies');
  if (cfg.lang) parts.push(`lang:${cfg.lang}`);
  return parts.join(' ');
}

async function extractTweets(page) {
  return page.$$eval('article[data-testid="tweet"]', (articles) => {
    return articles.map((article) => {
      const timeEl = article.querySelector('time');
      const datetime = timeEl ? timeEl.getAttribute('datetime') : null;
      const linkEl = timeEl ? timeEl.closest('a') : null;
      const url = linkEl ? linkEl.href : null;

      const textEl = article.querySelector('[data-testid="tweetText"]');
      const text = textEl ? textEl.innerText.replace(/\s+/g, ' ').trim() : '';

      let handle = null;
      let displayName = null;
      const userNameBlock = article.querySelector('[data-testid="User-Name"]');
      if (userNameBlock) {
        const links = userNameBlock.querySelectorAll('a[href^="/"]');
        for (const a of links) {
          const href = a.getAttribute('href') || '';
          if (/^\/[A-Za-z0-9_]+$/.test(href)) {
            handle = href.slice(1);
            break;
          }
        }
        displayName = (userNameBlock.innerText.split('\n')[0] || '').trim() || null;
      }

      const metrics = {};
      ['reply', 'retweet', 'like'].forEach((key) => {
        const el = article.querySelector(`[data-testid="${key}"]`);
        metrics[key] = el ? (el.getAttribute('aria-label') || el.textContent || '').trim() : null;
      });

      return { url, datetime, text, handle, displayName, metrics };
    });
  });
}

function toCsv(rows) {
  const headers = ['datetime', 'handle', 'displayName', 'text', 'url', 'reply', 'retweet', 'like'];
  const escape = (v) => {
    const s = v === null || v === undefined ? '' : String(v);
    return /[",\n]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
  };
  const lines = [headers.join(',')];
  for (const r of rows) {
    lines.push(
      headers
        .map((h) => (h === 'reply' || h === 'retweet' || h === 'like' ? r.metrics?.[h] : r[h]))
        .map(escape)
        .join(',')
    );
  }
  return lines.join('\n');
}

function matchesKeywords(text, cfg) {
  const lower = text.toLowerCase();
  const queryHit = lower.includes(cfg.query.toLowerCase());
  if (!cfg.strictKeywordFilter) return true;
  const kwHit = cfg.keywords.some((k) => lower.includes(k.toLowerCase()));
  return queryHit && kwHit;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const cfg = loadConfig(args);

  if (!fs.existsSync(STORAGE_STATE_PATH)) {
    console.error(
      `Brak pliku sesji: ${STORAGE_STATE_PATH}\n` +
        'Najpierw zaloguj sie jednorazowo lokalnie: npm run login'
    );
    process.exit(1);
  }
  if (!fs.existsSync(OUTPUT_DIR)) fs.mkdirSync(OUTPUT_DIR, { recursive: true });

  const cutoffMs = Date.now() - cfg.hoursBack * 60 * 60 * 1000;
  const searchQuery = buildSearchQuery(cfg);
  const url = `https://x.com/search?q=${encodeURIComponent(searchQuery)}&src=typed_query&f=live`;

  console.log(`Zapytanie wyszukiwania: ${searchQuery}`);
  console.log(`Okno czasowe: ostatnie ${cfg.hoursBack}h (od ${new Date(cutoffMs).toISOString()})`);

  const browser = await chromium.launch({ headless: cfg.headless });
  const context = await browser.newContext({
    storageState: STORAGE_STATE_PATH,
    viewport: { width: 1280, height: 1000 },
    userAgent:
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36',
  });
  const page = await context.newPage();

  await page.goto(url, { waitUntil: 'domcontentloaded' });

  if (page.url().includes('/login') || page.url().includes('/i/flow/login')) {
    console.error('Sesja wygasla lub jest nieprawidlowa - zaloguj sie ponownie: npm run login');
    await browser.close();
    process.exit(1);
  }

  try {
    await page.waitForSelector('article[data-testid="tweet"]', { timeout: 20000 });
  } catch {
    console.warn('Nie znaleziono zadnych postow dla tego zapytania (lub strona nie zaladowala sie poprawnie).');
  }

  const seen = new Map();
  let staleScrolls = 0;

  for (let i = 0; i < cfg.maxScrolls; i++) {
    const batch = await extractTweets(page);

    for (const t of batch) {
      if (!t.url || !t.datetime) continue;
      if (seen.has(t.url)) continue;
      if (!matchesKeywords(t.text, cfg)) continue;
      seen.set(t.url, t);
    }

    const times = batch.map((t) => t.datetime).filter(Boolean).map((d) => new Date(d).getTime());
    const oldestInBatch = times.length ? Math.min(...times) : null;

    if (oldestInBatch !== null && oldestInBatch < cutoffMs) {
      staleScrolls++;
    } else {
      staleScrolls = 0;
    }

    console.log(`Scroll ${i + 1}/${cfg.maxScrolls} - zebrano lacznie: ${seen.size}`);

    if (staleScrolls >= 3) {
      console.log('Kolejne posty sa juz starsze niz zadane okno czasowe - konczenie.');
      break;
    }
    if (seen.size >= cfg.maxTweets) {
      console.log('Osiagnieto limit maxTweets - konczenie.');
      break;
    }

    await page.mouse.wheel(0, 2200);
    await page.waitForTimeout(1200 + Math.random() * 1300);
  }

  await browser.close();

  const results = [...seen.values()]
    .filter((t) => new Date(t.datetime).getTime() >= cutoffMs)
    .sort((a, b) => new Date(b.datetime) - new Date(a.datetime));

  const stamp = new Date().toISOString().replace(/[:.]/g, '-');
  const jsonPath = path.join(OUTPUT_DIR, `x-posts-${stamp}.json`);
  const csvPath = path.join(OUTPUT_DIR, `x-posts-${stamp}.csv`);

  fs.writeFileSync(jsonPath, JSON.stringify(results, null, 2), 'utf8');
  fs.writeFileSync(csvPath, toCsv(results), 'utf8');

  console.log(`\nZnaleziono ${results.length} postow z ostatnich ${cfg.hoursBack}h.`);
  console.log(`Zapisano do:\n  ${jsonPath}\n  ${csvPath}`);
}

main().catch((err) => {
  console.error('Blad scrapera:', err);
  process.exit(1);
});
